### PR TITLE
Added exponential backoff

### PR DESCRIPTION
**Pull request by divyaac**
Link to ENT PR: https://github.com/hashicorp/vault-enterprise/pull/5444

closes https://github.com/hashicorp/vault/issues/12566
closes https://github.com/hashicorp/vault/issues/22789
closes https://github.com/hashicorp/vault/issues/24657

### CODE REVIEW

File: changelog/25497.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+```release-note:improvement
+agent: Use exponential backoff to sleep and then attempt a restart of the consul template server
```
Comment (VioletHynes) [inline changelog/25497.txt:2]:
```suggestion
```release-note:bug
agent: Fixed a bug where Agent would immediately restart the template server without backoff, potentially resulting in high load to Vault
```

**Comment by VioletHynes** [comment]
This is copied from the ent PR (that I think we can close):

I think we should use `sdk/backoff.go` for this, as this is the same backoff mechanism we use for auto-auth. It's familiar to customers in the way that it works, and that means we don't have multiple different 'kinds' of backoffs within Agent. We can use nice helpers like `NextSleep` that way, too!

Also, [we can use closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) to link the relevant GitHub issues, so that they get updated when this is merged, e.g.:

closes https://github.com/hashicorp/vault/issues/12566
closes https://github.com/hashicorp/vault/issues/22789
closes https://github.com/hashicorp/vault/issues/24657

File: command/agent/template/template.go [hunk 2]
```diff
@@ -191,6 +199,17 @@ func (ts *Server) Run(ctx context.Context, incoming chan string, templates []*ct
 				return fmt.Errorf("template server: %w", err)
 			}
 
+			// Calculate the amount of time to backoff using exponential backoff
+			sleep, err := restartBackoff.Next()
```
Comment (divyaac) [inline command/agent/template/template.go:203]: Using "next" here in stead of NextSleep so we can log how long it's sleeping for.

File: command/agent/exec/exec.go [hunk 3]
```diff
@@ -216,6 +223,21 @@ func (s *Server) Run(ctx context.Context, incomingVaultToken chan string) error
...
 
+			// Calculate the amount of time to backoff using exponential backoff
+			sleep, err := restartBackoff.Next()
+			if err != nil {
+				ts.logger.Error("template server: reached maximum number restart attempts")
+				restartBackoff.Reset()
+			}
+
+			// Sleep for the calculated backoff time then attempt to create a new runner
+			ts.logger.Warn(fmt.Sprintf("template server restart: retry attempt after %s", sleep))
```
Comment (VioletHynes) [inline command/agent/exec/exec.go:234]: I'm a little confused why this is repeated -- is this a copy/paste error?

**Review by VioletHynes — COMMENTED** [review]
Looks great! Just a small comment remaining.

File: command/agent/exec/exec.go (hunk 3 above)
Comment (divyaac) [reply command/agent/exec/exec.go:234]: Ooops yes copy and paste error. Removing now.

**Review by VioletHynes — APPROVED** [review]
Looks great! Thanks for refactoring to use `sdk/backoff`, and good spot fixing it in `exec.go` too. I can't really think of a good automated test for this functionality, but as long as you've manually tested, it looks good to go from my perspective!

Awesome work!

**Pull request merged**